sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/constraints/Constraint.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog.constraints;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.expressions.NamedReference;
+
+/**
+ * A constraint that restricts states of data in a table.
+ *
+ * @since 4.1.0
+ */
+@Evolving
+public interface Constraint {
+  /**
+   * Returns the name of this constraint.
+   */
+  String name();
+
+  /**
+   * Indicates whether this constraint is actively enforced. If enforced, data modifications
+   * that violate the constraint fail with a constraint violation error.
+   */
+  boolean enforced();
+
+  /**
+   * Indicates whether the existing data in the table satisfies this constraint. The constraint
+   * can be valid (the data is guaranteed to satisfy the constraint), invalid (some records violate
+   * the constraint), or unvalidated (the validity is unknown). The validation status is usually
+   * managed by the system and can't be modified by the user.
+   */
+  ValidationStatus validationStatus();
+
+  /**
+   * Indicates whether this constraint is assumed to hold true if the validity is unknown. Unlike
+   * the validation status, this flag is usually provided by the user as a hint to the system.
+   */
+  boolean rely();
+
+  /**
+   * Returns the definition of this constraint in the DDL format.
+   */
+  String toDDL();
+
+  /**
+   * Instantiates a builder for a CHECK constraint.
+   *
+   * @param name the constraint name
+   * @return a CHECK constraint builder
+   */
+  static Check.Builder check(String name) {
```

Comment:
How about `createCheckBuilder` ?

Reply:
I personally prefer shorter names whenever the usage/context is obvious enough. 

```
Constraint.check("con1").prediateSQL("id > 0").enforced(true).build();
```

This reads well to me and matches what we did for `ProcedureParameter`.

Reply:
+1 with @aokolnychyi , `check` is neat